src/lib.rs

```diff
@@ -19,55 +50,282 @@ pub mod common {
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
-pub struct Header {
-    pub sequence: u8,
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Default)]
+pub struct MavHeader {
     pub system_id: u8,
     pub component_id: u8,
+    pub sequence: u8,
+}
+
+
+/// Message framing marker for mavlink v1
+pub const MAV_STX: u8 = 0xFE;
+
+/// Message framing marker for mavlink v2
+pub const MAV_STX_V2: u8 = 0xFD;
+
+
+impl MavHeader {
+    /// Return a default GCS header, seq is replaced by the connector
+    /// so it can be ignored. Set `component_id` to your desired component ID.
+    pub fn get_default_header() -> MavHeader {
+        MavHeader {
+            system_id: 255,
+            component_id: 0,
+            sequence: 0,
+        }
+    }
+}
+
+/// Encapsulation of the Mavlink message and the header,
+/// important to preserve information about the sender system
+/// and component id
+#[derive(Debug, Clone)]
+pub struct MavFrame {
+    pub header: MavHeader,
+    pub msg: MavMessage,
+}
+
+impl MavFrame {
+    /// Create a new frame with given message
+    pub fn new(msg: MavMessage) -> MavFrame {
+        MavFrame {
+            header: MavHeader::get_default_header(),
+            msg
+        }
+    }
+
+    /// Serialize frame into a vector, so it can be send
+    /// over a socket for example
+    pub fn ser(&self) -> Vec<u8> {
+        let mut v = vec![];
+
+        // serialize header
+        v.push(self.header.system_id);
+        v.push(self.header.component_id);
+        v.push(self.header.sequence);
+
+        // message id
+        #[cfg(feature="mavlink2")]
+        {
+            let bytes: [u8; 4] = unsafe { transmute(self.msg.message_id().to_le()) };
+            v.extend_from_slice(&bytes);
+        }
+        #[cfg(not(feature="mavlink2"))]
+        v.push(self.msg.message_id());
+
+        // serialize message
+        v.append(&mut self.msg.ser());
+
+        v
+    }
+
+    /// Deserialize MavFrame from a slice that has been received from
+    /// for example a socket.
+    pub fn deser(input: &[u8]) -> Option<Self> {
+        let mut buf = Bytes::from(input).into_buf();
+
+        let system_id = buf.get_u8();
+        let component_id = buf.get_u8();
+        let sequence = buf.get_u8();
+        let header = MavHeader{system_id,component_id,sequence};
+
+        #[cfg(not(feature="mavlink2"))]
+        let msg_id = buf.get_u8();
+
+        #[cfg(feature="mavlink2")]
+        let msg_id = buf.get_u32_le();
+
+        if let Some(msg) = MavMessage::parse(msg_id, &buf.collect::<Vec<u8>>()) {
+            Some(MavFrame {header, msg})
+        } else {
+            None
+        }
+    }
+
+    /// Return the frame header
+    pub fn header(&self) -> MavHeader {
+        self.header
+    }
 }
 
-/// Read a MAVLink message from a Read stream.
-pub fn read<R: Read>(r: &mut R) -> io::Result<(Header, MavMessage)> {
+/// Read a MAVLink v1  message from a Read stream.
+#[cfg(all(feature = "std", not(feature="mavlink2")))]
```

Comment:
The `mavlink2` cfg flag shouldn't remove runtime support for Mavlink 1. Cargo feature flags have to be purely additive because if multiple dependencies of a program transitively depend on this one, it gets compiled with the union of their cargo features.

Reply:
The comment about additive features is fair.  @podhrmic we should talk about a strategy for supporting both mavlink1 and mavlink2.  Other implementations support either one or the other... I think at least one implementation supports negotiation (or fallback to v1).